lib/tabulation/src/Data/Tabulation.hs

```diff
@@ -18,4 +19,134 @@ class HasFields a where
+  XYHKDField_X :: XYHKDField (f ())
+  XYHKDField_Y :: XYHKDField (f Bool)
+
+instance HasFields (XYHKD f) where
+  type Field (XYHKD f) = XYHKDField
+  fieldLens = \\case
+    XYHKDField_X -> \\f (XYHKD x y) -> flip XYHKD y '<$>' f x
+    XYHKDField_Y -> lens _y' $ \\xy y -> xy { _y' = y }
+  tabulateFieldsA g = pure XYHKD
+    '<*>' g XYHKDField_X
+    '<*>' g XYHKDField_Y
+@
+
+While `tabulateFieldsA` compiles, `fieldLens` doesn't, with the same sort of error in both its cases.
+
+@
+   • Could not deduce: f2 ~ f
+      from the context: x ~ f2 ()
+        bound by a pattern with constructor:
+                   XYHKDField_X :: forall (f :: * -> *). XYHKDField (f ()),
+                 in a case alternative
+        at ../lib/tabulation/src/Data/Tabulation.hs:68:5-16
+      ‘f2’ is a rigid type variable bound by
+        a pattern with constructor:
+          XYHKDField_X :: forall (f :: * -> *). XYHKDField (f ()),
+        in a case alternative
+        at ../lib/tabulation/src/Data/Tabulation.hs:68:5-16
+      ‘f’ is a rigid type variable bound by
+        the instance declaration
+        at ../lib/tabulation/src/Data/Tabulation.hs:65:10-28
+      Expected type: x
+        Actual type: f ()
+@
+
+The `XYHKDField` constructors must work for all `f`, but the one from `XYHKD` is a specific one (it's quantified outside), so this direction doesn't work, but the one in `tabulateFieldsA` does.
+
+We can fix this by threading the `f` from the record to the associated GADT.
+
+@
+data XYHKDField f a where
+  XYHKDField_X :: XYHKDField f (f ())
+  XYHKDField_Y :: XYHKDField f (f Bool)
+
+instance HasFields (XYHKD f) where
+  type Field (XYHKD f) = XYHKDField f
+@
+
+but that's cumbersome and takes some work to figure out once you first try to use `Field` with HKD.
+
+To solve these problems, one can use `FieldHKD` instead.
+
+@
+instance HasFields (XYHKD f) where
+  type Field (XYHKD f) = FieldHKD XYField f
+  fieldLens = \\case
+    FieldHKD XYField_X -> \\f (XYHKD x y) -> flip XYHKD y '<$>' f x
+    FieldHKD XYField_Y -> lens _y' $ \\xy y -> xy { _y' = y }
+  tabulateFieldsA g = pure XYHKD
+    '<*>' g (FieldHKD XYField_X)
+    '<*>' g (FieldHKD XYField_Y)
+@
+
+This also lets us capture the notion that the fields of the HKD version are related to the base version by a `f` layer by transforming one into the other.
+
+@
+liftKind
+  :: ('HasFields' t, 'HasFields' (t' f), 'Field' (t' f) ~ FieldHKD ('Field' t) f)
+  => (forall x. x -> f x) -> t -> t' f
+liftKind f r = 'tabulateFields' $ \(FieldHKD field) -> f ('indexField' r field)
+
+unliftKind
+  :: ('HasFields' t, 'HasFields' (t' f), 'Field' (t' f) ~ FieldHKD ('Field' t) f)
+  => (forall x. f x -> x) -> t' f -> t
+unliftKind f r = 'tabulateFields' $ \field -> f ('indexField' r (FieldHKD field))
+
+xyPure :: 'Applicative' f => XY -> XYHKD f
+xyPure = 'liftKind' 'pure'
+
+xyRunIdentity :: XYHKD 'Identity' -> XY
+xyRunIdentity = 'unliftKind' 'runIdentity'
+@
+-}
+data FieldHKD field f x where
```

Comment:
Can this be a newtype? If so, can lift/unliftkind be coercions?

Reply:
Yeah this has no existential so it should be fine as a newtype.

Reply:
Think I'm a bit lost here
1. how would I enforce the contents/index relationship if a newtype [doesn't allow GADTs](https://gitlab.haskell.org/ghc/ghc/-/issues/4009)? 
   ```haskell
   newtype FieldHKD field f (f x) = FieldHKD (field x)
   ```
2. liftKind/unliftKind don't convert between fields, they convert between the records, which are not newtypes of eachother. I can only imagine that being possible with [Unsaturated Type Families](https://github.com/ghc-proposals/ghc-proposals/pull/242) and something like
   ```haskell
   type family Id a where
     Id a = a

   data XYHKD f = XYHKD
     { _x' :: f ()
     , _y' :: f Bool
     }

   newtype XY = XY (XYHKD Id)
   ```

Reply:
Ah indeed
```haskell
newtype FieldHKD field f x where
  FieldHKD :: field x -> FieldHKD field f (f x)
```
doesn't quite work because there is a secret coersion for the `(f x)`

Reply:
Trying to use `GADTSyntax` with `newtype` confirms this isn't doable:

```• A newtype constructor cannot have existential type variables```